src/mqtt.service.ts

```diff
@@ -1,13 +1,17 @@
-import { Inject, Injectable } from '@nestjs/common';
+import { Inject, Injectable, OnModuleDestroy } from '@nestjs/common';
 import { MQTT_CLIENT_INSTANCE } from './mqtt.constants';
 import { Client, Packet, IClientPublishOptions, IClientSubscribeOptions, ISubscriptionGrant } from 'mqtt';
 
 @Injectable()
-export class MqttService {
+export class MqttService implements OnModuleDestroy {
   constructor(
     @Inject(MQTT_CLIENT_INSTANCE) private readonly client: Client,
   ) {}
 
+  onModuleDestroy() {
```

Comment:
This requires adding `app.enableShutdownHooks()` to `main.ts`.

IMHO, the proper way to deal with the closing the client(s) is doing it outside of `nest-mqtt`:
- `nest-mqtt` should export (if it does not already does so) the `client` object;
- users should add `app.enableShutdownHooks()` to their `main.ts`;
- create a class for `OnModuleDestroy()`;
- add that class to the `providers` array in `app.module.ts`.

I don’t really think there is anything to do in `nest-mqtt` (unless `client` object is not yet exported). However, I don’t use `net-mqtt` anymore, as I needed to many stuff to implement in order to make it work as I need it to work, therefore I re-implemented it within my codebase.

Reply:
Looks like [it's not exporting the client yet](https://github.com/microud/nest-mqtt/blob/master/src/mqtt.module.ts#L22).

I agree that this line is kind of misplaced. Also delegating this responsibility to users is more flexible in case you don't want `client#end` to be called for whatever reason.

I'll submit a second PR exposing the `client` object and documenting how to properly close it's connection. This way, whichever approach is decided to be the best can be merged and the other closed.